pgxn/neon/libpagestore.c

```diff
@@ -64,87 +55,176 @@ int			flush_every_n_requests = 8;
 
 int			n_reconnect_attempts = 0;
 int			max_reconnect_attempts = 60;
+int			stripe_size;
+
+bool	(*old_redo_read_buffer_filter) (XLogReaderState *record, uint8 block_id) = NULL;
 
-#define MAX_PAGESERVER_CONNSTRING_SIZE 256
+static bool pageserver_flush(shardno_t shard_no);
+static void pageserver_disconnect(shardno_t shard_no);
+static void AssignPageserverConnstring(const char *newval, void *extra);
+
+static shmem_startup_hook_type prev_shmem_startup_hook;
+#if PG_VERSION_NUM>=150000
+static shmem_request_hook_type prev_shmem_request_hook;
+#endif
 
 typedef struct
```

Comment:
Please add comment explaining what this is. Something like:

```
+/*
+ * ShardMap is kept in shared memory. It contains the connection strings for
+ * each shard.
+ *
+ * XXX: Explain the update protocol with the counters. Why now an LWLock?
+ */
```

Reply:
I don't understand why these locking changes are part of this PR. If we need to switch to separate begin- and end counters, let's do that as a separate PR.

Reply:
> Please add comment explaining what this is. Something like:

Done

> I don't understand why these locking changes are part of this PR. If we need to switch to separate begin- and end counters, let's do that as a separate PR.

Sorry, this PR was implemented before @save-buffer 's  PR for support of dynamic change of PS connection string. This is why it is implemented it in different way.
The status quo is the following: current implementation in main is not correct, it doesn't address the problem that postmaster can not wait. We discussed it with  @save-buffer  and come to the conclusion that it needs to be fixed because it can cause crash of postmaster (with very small probability, if config is reloaded at the moment when some backend is holding lock). But if this sharding PR is going to be merged soon, then may be it is not needed.